### DISCUSSION TITLE
can create_kb just build once when train the entity linking model

### DISCUSSION THREAD

**Discussion by Zhou0418** [explosion/spaCy#7717] [Help: Coding & Implementations] [answered]
## Your Environment

- Operating System:
- Python Version Used:
- spaCy Version Used:
- Environment Information:

**Comment by svlandeg** [comment]
Hi! As a maintenance note - this type of question is better suited for the discussion forum, so I'll move it there. The original issue will be closed, but will link through to the new, open discussion thread.

**Comment by svlandeg** [comment]
@Zhou0418 : can you elaborate a little on your question?

**Comment by Zhou0418** [comment, marked as answer]
> 
![image](https://user-images.githubusercontent.com/29883149/114155479-1f845180-9954-11eb-972a-94b8b7796670.png)
In spacy v2 we can create kb and save it.
But in spacy v3 we must put it with entity_linker train together as a function and can't save KB only
it will be called every time when i train el model and cost much time to repeat the same thing!

**Reply by svlandeg** [reply]
I see what you mean, but you can definitely avoid creating the KB multiple times from scratch:
- You can create a KB and save it to file with [`kb.to_disk`](https://spacy.io/api/kb#to_disk)
- You can read the KB back in with [`kb.from_disk`](https://spacy.io/api/kb#from_disk)
- Better yet, you can use the [config system](https://spacy.io/usage/training#config) to specify the KB generation function, and use for instance [`spacy.KBFromFile.v1`](https://spacy.io/api/architectures#KBFromFile) as showcased by [this example project](https://github.com/explosion/projects/tree/v3/tutorials/nel_emerson):
```
[initialize.components.entity_linker]

[initialize.components.entity_linker.kb_loader]
@misc = "spacy.KBFromFile.v1"
kb_path = ${paths.kb}
``` 
- In the same `initialize` block, you can also use your custom `create_kb` function (after you [register](https://spacy.io/api/top-level#registry) it) - it will still only be called one (cf next point)
- Once you've serialized the whole pipeline, e.g. by calling `nlp.to_disk` or when running the CLI training and creating output model directories, the `entity_linker` pipeline will [serialize](s) its internal KB within the same directory that the model weights are stored. This means that you'll be able to just load your `nlp` object from file, and the `entity_linker` pipe` will load the KB from its subdirectory. The KB creation function (as the one you showed) is not called at that moment anymore.

**Reply by Zhou0418** [reply]
i have done this like below
![image](https://user-images.githubusercontent.com/29883149/114330287-f510df00-9b73-11eb-8d00-ec90e460da0f.png)

and often erro like this
![image](https://user-images.githubusercontent.com/29883149/114330212-cabf2180-9b73-11eb-87c9-4fcdb0457d28.png)

**Reply by svlandeg** [reply]
As the error message states, you have to provide a function into `entity_linker.set_kb`, and not the actual KB. Like you had it in your original question. This function will automatically be called with the `vocab` instance of the pipeline that the `entity_linker` is a part of.

Are you working with the config system in v3? Then you can use the `initialize` block as I described before.
If you're programming a custom training loop in v3, you can do something like this:
```
entity_linker = nlp.add_pipe("entity_linker")
entity_linker.initialize(lambda: [], nlp=nlp, kb_loader=create_kb)
```

(by the way - for future reference - it's easier for us if you copy/paste the code snippets and mark them up with code blocks, instead of screen shots of code)

**Reply by Zhou0418** [reply]
So i can't load KB path directly for training entity_linking model except using config system right? 

`entity_linker.initialize(lambda: [], nlp=nlp, kb_loader=create_kb)`

this code it's still call the create_kb function.
Next step i will choose to use config system! and for future reference i will use code blocks, Thanks for your patience!!!~

**Reply by svlandeg** [reply]
So yes, this `entity_linker.initialize` function calls the `create_kb` function - but it's important to understand that this will happen only ONCE. `create_kb` will create the KB and then the `entity_linker` component will serialize and deserialize this KB from file, it won't rerun the `create_kb` function.

This use-case is exactly why we have the "initialize" functionality: because sometimes you want to create something once, and then afterwards rely on the components to keep their internal data consistent.

**Reply by Zhou0418** [reply]
I got it! Thanks very much~

**Comment by svlandeg** [comment]
Just want to add to this thread, that we've updated [this notebook](https://github.com/explosion/projects/blob/v3/tutorials/nel_emerson/notebooks/notebook_video.ipynb) with the full run-through on how to create and train an entity linker in spaCy v3. It also shows how to initialize the `entity_linker` with `kb_loader` argument.
